### PR TITLE
Compare release candidates and milestones in addition to major.minor.patch

### PR DESCRIPTION
**Pull request by tgodzik**
This showed up as a warning when using 3.0.0-M1:

image

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/ScalaVersions.scala [hunk 1]
```diff
@@ -60,15 +60,15 @@ object ScalaVersions {
     latestBinaryVersionFor(scalaVersion)
       .map(latest =>
         latest != scalaVersion && SemVer
-          .isCompatibleVersion(latest, scalaVersion)
+          .isLaterVersion(latest, scalaVersion)
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/ScalaVersions.scala:63]: This was the main issue. `3.0.0-M1` was equal to `3.0.0-M2`, because we were comparing only major/minor/patch. 
First of all it should not say it's a future version if it's equal, second of all it would be good to support also milestones/release candidates now.

File: tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala [hunk 2]
```diff
@@ -143,4 +143,59 @@ class ScalaVersionsSuite extends BaseSuite {
...
+    assert(
+      SemVer.isCompatibleVersion("3.0.0-M1", "3.0.0-M2")
+    )
+  }
+
+  test("compare-milestones-M1<M2") {
+    assert(
+      !SemVer.isCompatibleVersion("3.0.0-M2", "3.0.0-M1")
+    )
+  }
```
Comment (ckipp01) [inline tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala:157]: These two tests will always be the same no? If one fails, the other should fail. If on passes the other should pass.

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
+
 object SemVer {
 
-  case class Version(major: Int, minor: Int, patch: Int) {
+  case class Version(
+      major: Int,
+      minor: Int,
+      patch: Int,
+      releaseCandidate: Int,
+      milestone: Int
```
Comment (ckipp01) [inline mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:12]: Do you think it's best to model this as an `Int`? Would an `Option[Int]` not be more appropriate?

Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:12]: I tried with option but that became rather complex, since I had to include the scenario where we compared release candidate with milestone. Though this will not work currently, so I might need to rethink it and do it with option.

File: tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala (hunk 2 above)
Comment (tgodzik) [reply tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala:157]: I also switched the order of the parameters, so this are not strictly opposite tests. At least I have seen one fail and the other not 😅

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala (hunk 3 above)
Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:12]: 3.0.0 will be lower than 3.0.0-M1 currently, which I didn't think of before 😱

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala [hunk 4]
```diff
@@ -1,12 +1,22 @@
...
+  ) {
     def >(that: Version): Boolean = {
       this.major > that.major ||
       (this.major == that.major && this.minor > that.minor) ||
-      (this.major == that.major && this.minor == that.minor && this.patch > that.patch)
+      (this.major == that.major && this.minor == that.minor && this.patch > that.patch) ||
+      (this.major == that.major && this.minor == that.minor && this.patch == that.patch && this.releaseCandidate > that.releaseCandidate) ||
+      (this.major == that.major && this.minor == that.minor && this.patch == that.patch && this.releaseCandidate == that.releaseCandidate && this.milestone > that.milestone)
```
Comment (tanishiking) [inline mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:19]: As you notice, we have to compare the Pre-Release version in a different way.

FYI, here's the logic of comparing semantic versions in scalafmt
https://github.com/scalameta/scalafmt/blob/f05136cdd5ceeeabdb6054658d315f7a0a3d025e/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala#L15-L18


Following the logic from scalafmt, these two lines would be something like this (if we use `0` for no pre-release version):

```scala
(this.major == that.major && this.minor == that.minor && this.patch == that.patch && isLaterPreRelease(this.releaseCandidate, that.releaseCandidate)) ||
(this.major == that.major && this.minor == that.minor && this.patch == that.patch && isLaterPreRelease(this.releaseCandidate, that.releaseCandidate)) && isLaterPreRelease(this.milestone, that.milestone)

// returns true if v1 is later than v2 (for pre-release version)
def isLaterPreRelease(v1: Int, v2: Int): Boolean {
  v1 == 0 || (v2 !== 0 && v1 > v2)
}
```

Comparing two semantic versions is always complexed 😅

Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala]: Thanks for the hint! I went with the Option solution suggested by @ckipp01

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala (hunk 3 above)
Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala]: Added option, aaaaand a lot more tests haha.

File: tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala (hunk 2 above)
Comment (tgodzik) [reply tests/unit/src/test/scala/tests/ScalaVersionsSuite.scala]: Ach! I know how it's called. I am actually testing transitivity!

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala [hunk 5]
```diff
@@ -1,27 +1,71 @@
...
       this.major > that.major ||
       (this.major == that.major && this.minor > that.minor) ||
-      (this.major == that.major && this.minor == that.minor && this.patch > that.patch)
+      (this.major == that.major && this.minor == that.minor && this.patch > that.patch) ||
+      // 3.0.0 > 3.0.0-M2 and 3.0.0 > 3.0.0-RC1
+      (this.milestone.isEmpty && this.releaseCandidate.isEmpty && (that.milestone.isDefined || that.releaseCandidate.isDefined)) ||
+      // 3.0.0-RC2 > 3.0.0-RC1
+      comparePreRelease(that, (v: Version) => v.releaseCandidate) ||
+      // 3.0.0-M2 > 3.0.0-M1
+      comparePreRelease(that, (v: Version) => v.milestone)
```
Comment (tanishiking) [inline mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:23]: Sorry, I think I gave a wrong hint...
How about comparing versions with different pre-release versions? The following test should pass, but I guess it fails now.
 
```
    assert(
      SemVer.isLaterScalaVersion("3.0.0-M1", "3.0.0-RC1")
    )
```

We have to compare `releaseCandidate` and `milestone` simultaneously.
Or store the label (like `RC` or `M` part) to `Version` and compare those labels by ASCII sort order (this would be more generic implementation).

```
  case class Version(
      major: Int,
      minor: Int,
      patch: Int,
      preRelease: Option[(String, Int)], // something like this?
  ) 
```

> Identifiers with letters or hyphens are compared lexically in ASCII sort order.
> https://semver.org/#spec-item-11

Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:23]: ach, righ!. Every time I think I got it right, there are still edge cases. I will add tests for that too and try figure it out. Good catch.

Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:25]: Added additional check:
```scala
      // 3.0.0-RC1 > 3.0.0-M1
      this.releaseCandidate.isDefined && that.milestone.isDefined ||
```

**Review by tanishiking — APPROVED** [review]
LGTM! :+1:

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala (hunk 5 above)
Comment (ckipp01) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:25]: lol so many checks. Great job on figuring these out.

**Review by ckipp01 — APPROVED** [review]
Looks good, and thanks for adding in all the tests! LGTM!

File: mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala (hunk 5 above)
Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala:25]: Yeah, that's why I was not keen on options at first. But I think with the comments it's not that bad. And I added a lot of tests :sweat_smile:
